File: python/14_Longest-Common-Prefix.py
class Solution:
    def longestCommonPrefix(self, strs):
        #Only one case
        if len(strs) == 1:
            return strs[0]
        #Check empty str
        for ele in strs:
            if len(ele) == 0:
                return ""
        pivot = 0
        for idx, item in enumerate(zip(*strs)):
            itemInSet = set(item)
            pivot = idx
            #print(item)
            if len(itemInSet) != 1:
                #print("idx = ", idx)
                return strs[0][:pivot]
        return strs[0][:pivot+1]

        
if __name__ == "__main__":
    s = Solution()
    strs = ["flower","flow","flight"]
    print("ansStr = ", s.longestCommonPrefix(strs))

    strs = ["dog","racecar","car"]  
    print("ansStr = ", s.longestCommonPrefix(strs))

    strs = ["ab", "a"]
    print("ansStr = ", s.longestCommonPrefix(strs))

    strs = ["abab","aba",""]
    print("ansStr = ", s.longestCommonPrefix(strs))

File: c++/14_Longest-Common-Prefix.cpp
class Solution {
public:
    string longestCommonPrefix(vector<string>& strs) {

        for(int char_idx = 0; char_idx < strs[0].size(); ++char_idx){
            char c = strs[0][char_idx];
            for(int i = 0; i < strs.size(); ++i){
                if(char_idx == strs[i].length()) return strs[i];

                if(strs[i][char_idx] != c){
                    return strs[0].substr(0, char_idx);
                }
            }
        }

        // The entire strs[0] is common prefix
        return strs[0];
    }
};